### ISSUE TITLE
Exception thrown by MatrixStorage.GetHashCode for some matrices where ColumnCount > RowCount

### ISSUE THREAD

**Issue by mjmckp** [mathnet/mathnet-numerics#534]
The `MatrixStorage.GetHashCode` method https://github.com/mathnet/mathnet-numerics/blob/532f7439075d51a726071f45202b6c4a8a508089/src/Numerics/LinearAlgebra/Storage/MatrixStorage.cs#L185 throws an `IndexOutOfRangeException` in some cases.  For example, the following F# script
```
    #I @"C:\packages\MathNet.Numerics.3.20.0\lib\net40\"
    #I @"C:\packages\MathNet.Numerics.FSharp.3.20.0\lib\net40\"
    #r "MathNet.Numerics.dll"
    #r "MathNet.Numerics.FSharp.dll"

    open MathNet.Numerics.LinearAlgebra
    open MathNet.Numerics.LinearAlgebra.Double

    for rows=1 to 10 do
        for cols=1 to 10 do
            let m : Matrix<float> = DenseMatrix.zero rows cols
            try m.GetHashCode() |> ignore
            with e ->
                printfn "GetHashCode failed for %d x %d matrix (%s)" rows cols e.Message
```
prints out the following to the console:

```
GetHashCode failed for 2 x 4 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 2 x 5 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 2 x 6 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 2 x 7 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 2 x 8 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 2 x 9 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 2 x 10 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 3 x 5 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 3 x 6 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 3 x 7 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 3 x 8 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 3 x 9 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 3 x 10 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 4 x 6 matrix (Index was outside the bounds of the array.)
GetHashCode failed for 4 x 8 matrix (Index was outside the bounds of the array.)
```

The issue appears to be the line `var row = (i - col)/RowCount;`, which seems like it should instead be `var row = (i - col) % RowCount;`.   I'm not sure why `i-col` is used instead of just `i`, though.  I will submit a PR with this change.

**Comment by cdrnet** [comment]
Thanks for the analysis and the fix. Yes, we certainly need to constrain by RowCount there.

I agree that even with the fix the algorithm doesn't make sense. For example, for a 3x3 matrix, 9 values are used, but in fact only values from the top row are considered.

It seems to me, the intended behavior would actually be `var row = i/ColumnCount;`. In this case it would simply go through the first 25 fields in a simple order, row by row.

**Comment by cdrnet** [comment]
Released in v4.0.0-alpha04.

Let me know if there is a need to backport this to v3 as well.
